Data/ByteString/Lazy/ReadNat.hs

```diff
@@ -0,0 +1,209 @@
+        = loop cs valid cnt acc ns
+    loop s@(Chunk c@(BI.BS _ !len) cs) !valid !cnt !acc ns
+        = case natdigits c cnt acc ns of
+            Result 0 _ _ _ -- no more digits found
+                | valid
+                  -> let !r = kont (combine cnt acc ns) in Just (r, s)
+                | otherwise
+                  -> Nothing
+            Result used d a ms
+                | used < len -- input not entirely digits
+                  -> let !c' = BU.unsafeDrop used c
+                         s'  = Chunk c' cs
+                         !r = kont (combine d a ms)
+                      in Just (r, s')
+                | otherwise -- perhaps more digits from the remaining chunks
+                  -> loop cs True d a ms
+#endif
+
+----- Misc functions
+
+-- | Largest decimal digit count that never overflows the accumulator
+--
+safeLog :: Word
+#if WORD_SIZE_IN_BITS == 32
+safeLog = 9
+#elif WORD_SIZE_IN_BITS == 64
+safeLog = 19
+#endif
+
+-- | 10-power base for little-ending sequence of ~Word-sized "digits"
+-- For n*log(n) cost, these are recursively merged pairwise, while
+-- squaring the base.
+safeBase :: Natural
+safeBase = 10 ^ safeLog
+
+-- The intermediate representation is a little-endian sequence in base @10^19@,
+-- prefixed by an initial element in base @10^cnt@ for some `cnt` between 1 and
+-- 19.  The final result is obtained by recursive pairwise merging of the tail
+-- followed by a final multiplication by 10^cnt and addition of the head.
+--
+natdigits :: BI.ByteString -- ^ Input chunk
+          -> Word         -- ^ partial digit count
+          -> Word         -- ^ accumulated element
+          -> [Natural]    -- ^ accumulated MSB elements
+          -> Result
+{-# INLINE natdigits #-}
+natdigits (BI.BS !fp !len) = \ cnt acc ns ->
+    BI.accursedUnutterablePerformIO $
+        BI.unsafeWithForeignPtr fp $ \ ptr -> do
+            let end = ptr `plusPtr` len
+            go ptr end cnt acc ns ptr
+  where
+    go start end = loop
+      where
+        loop :: Word -> Word -> [Natural] -> Ptr Word8 -> IO Result
+        loop !cnt !acc ns !ptr
+            | ptr == end = return $ Result (ptr `minusPtr` start) cnt acc ns
+            | otherwise = fromDigit <$> peek ptr >>= \ !d ->
+                if d > 9
+                then return $ Result (ptr `minusPtr` start) cnt acc ns
+                else if cnt < safeLog
+                then loop (cnt+1) (10*acc + d) ns $ ptr `plusPtr` 1
+                else loop 1 d (toNatural acc : ns) $ ptr `plusPtr` 1
+
+-- | Merge least-significant head with reduction of of little-endian tail.
+combine :: Word             -- ^ count of digits in head element
+        -> Word             -- ^ value of head element
+        -> [Natural]        -- ^ tail elements (base @10^'safeLog'@)
+        -> Natural
+{-# INLINE combine #-}
+combine !_   !acc [] = toNatural acc
+combine !cnt !acc ns = 10^cnt * combine1 safeBase ns + toNatural acc
```

Comment:
`10 ^ cnt` might be better computed in `Word`.

Reply:
I see, and then converted to Natural, makes sense, coming up...